File: binutils/gdb/testsuite/gdb.python/py-pp-maint.py
# Copyright (C) 2010-2024 Free Software Foundation, Inc.

# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

# This file is part of the GDB testsuite.  It tests python pretty
# printers.

import re
import gdb.types
import gdb.printing


def lookup_function_lookup_test(val):
    class PrintFunctionLookup(object):
        def __init__(self, val):
            self.val = val

        def to_string(self):
            return "x=<" + str(self.val["x"]) + "> y=<" + str(self.val["y"]) + ">"

    typename = gdb.types.get_basic_type(val.type).tag
    # Note: typename could be None.
    if typename == "function_lookup_test":
        return PrintFunctionLookup(val)
    return None


class pp_s(object):
    def __init__(self, val):
        self.val = val

    def to_string(self):
        a = self.val["a"]
        b = self.val["b"]
        if a.address != b:
            raise Exception("&a(%s) != b(%s)" % (str(a.address), str(b)))
        return "a=<" + str(self.val["a"]) + "> b=<" + str(self.val["b"]) + ">"


class pp_ss(object):
    def __init__(self, val):
        self.val = val

    def to_string(self):
        return "a=<" + str(self.val["a"]) + "> b=<" + str(self.val["b"]) + ">"


def build_pretty_printer():
    pp = gdb.printing.RegexpCollectionPrettyPrinter("pp-test")

    pp.add_printer("struct s", "^struct s$", pp_s)
    pp.add_printer("s", "^s$", pp_s)

    # Use a lambda this time to exercise doing things this way.
    pp.add_printer("struct ss", "^struct ss$", lambda val: pp_ss(val))
    pp.add_printer("ss", "^ss$", lambda val: pp_ss(val))

    pp.add_printer(
        "enum flag_enum",
        "^flag_enum$",
        gdb.printing.FlagEnumerationPrinter("enum flag_enum"),
    )

    return pp


gdb.printing.register_pretty_printer(gdb, lookup_function_lookup_test)
my_pretty_printer = build_pretty_printer()
gdb.printing.register_pretty_printer(gdb, my_pretty_printer)

File: binutils/gdb/testsuite/gdb.python/py-pp-maint.c
/* This testcase is part of GDB, the GNU debugger.

   Copyright 2010-2024 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 3 of the License, or
   (at your option) any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.  */

#include <string.h>


enum flag_enum
  {
    /* Define the enumeration values in an unsorted manner to verify that we
       effectively sort them by value.  */
    FOO_MASK = 0x07,
    FOO_1    = 0x01,
    FOO_2    = 0x02,
    FOO_3    = 0x04,

    BAR_MASK = 0x70,
    BAR_1    = 0x10,
    BAR_2    = 0x20,
    BAR_3    = 0x40,
  };

enum flag_enum fval;

struct function_lookup_test
{
  int x,y;
};

void
init_flt (struct function_lookup_test *p, int x, int y)
{
  p->x = x;
  p->y = y;
}

struct s
{
  int a;
  int *b;
};

struct ss
{
  struct s a;
  struct s b;
};

void
init_s (struct s *s, int a)
{
  s->a = a;
  s->b = &s->a;
}

void
init_ss (struct ss *s, int a, int b)
{
  init_s (&s->a, a);
  init_s (&s->b, b);
}

int
main ()
{
  struct function_lookup_test flt;
  struct ss ss;

  init_flt (&flt, 42, 43);
  init_ss (&ss, 1, 2);
  
  return 0;      /* break to inspect */
}
